### PR TITLE
refactor ml for more flexibility, include extreme gradient boosting

### PR DESCRIPTION
**Pull request by LokiLuciferase**
I was interested to see how XGBoost fares in this task compared to linear SVMs.

## What was done
- create abstract TrexClassifier to facilitate different classifier algorithms
- create TrexSVM inheriting from TrexClassifier, implementing PICASVM functionality
- create TrexXGB which implements extreme gradient boosting as alternative classifier

Thanks to nice separation of concerns and usage of pipelines this was not a lot of work 😍 

## Results of Benchmarks
Runtime of both algs is roughly comparable since the base training of XGB is slower, but in turn no feature elimination has to be performed (is done internally by XGB). Also, XGBoost classifiers can output probabilities so no need to calibrate it.

Initial tests with the datasets I had lying around show some increased bACC at optimal comple/conta and reduced inter-fold variance, however performance at (very) low completeness is reduced. 

T3SS/SVM
t3ss_svm

T3SS/XGB
t3ss_xgb

Psychro_or_tolerant/SVM
psychro_or_tolerant_svm

Psychro_or_tolerant/XGB
psychro_or_tolerant_xgb

Sulfate_reducer/SVM
sulfate_reducer_svm

Sulfate_reducer/XGB
sulfate_reducer_xgb

Aerobe_old/SVM
Aerobe_old_svm

Aerobe_old/XGB
Aerobe_old_xgb

### CODE REVIEW

File: pica/ml/classifiers/svm.py [hunk 1]
```diff
@@ -4,29 +4,103 @@
...
 from pica.ml.cccv import CompleContaCV
+from pica.ml.trex_classifier import TrexClassifier
 from pica.util.logging import get_logger
 from pica.util.helpers import get_x_y_tn, get_groups
 from pica.ml.feature_select import recursive_feature_elimination, compress_vocabulary, DEFAULT_STEP_SIZE,\
     DEFAULT_SCORING_FUNCTION
 
 
+class TrexSVM(TrexClassifier):
+    def __init__(self, C: float = 5, penalty: str = "l2", tol: float = 1,
```
Comment (VarIr) [inline pica/ml/classifiers/svm.py]: If we annotate values as float, it'd be nice to initialize them as float as well,
e.g. `C: float = 5.` and `tol: float = 1.`
This applies to all classes here.

File: pica/ml/classifiers/xgbm.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Class which encapsulates a sklearn Pipeline of CountVectorizer (for vectorization of features) and
+    xgb.sklearn.GradientBoostingClassifier.
+    Provides train() and crossvalidate() functionality equivalent to train.py and crossvalidateMT.py.
+
+    :param random_state: A integer randomness seed for a Mersienne Twister (see np.random.RandomState)
+    :param kwargs: Any additional named arguments are passed to the XGBClassifier constructor.
+    """
+    def __init__(self, max_depth: int = 4, learning_rate: float = 0.05,
+                 n_estimators: int = 30, gamma: float = 0, min_child_weight: int = 1,
```
Comment (VarIr) [inline pica/ml/classifiers/xgbm.py]: `gamma: float = 0.`

**Review by VarIr — COMMENTED** [review]
Thanks for the code, cool idea!
At a first quick glance LGTM, with just a minor comment about default values in constructors.
Do you think test coverage could be increased?

File: pica/ml/classifiers/svm.py (hunk 1 above)
Comment (LokiLuciferase) [reply pica/ml/classifiers/svm.py]: Yeah, correct dtype make sense, definitely.

Sorry to say I haven't looked into the test suite at all, but I surely can take some inspiration from `test_svm.py` to write analogous tests for `TrexClassifier`s.

**Comment by LokiLuciferase** [comment]
I added a test class for `TrexClassifier`s, but there's yet no data for XGB models (have to compute this at some point) to validate output - when I add that, coverage should get another bump. However, the TrexSVM classifier passes all output validations, showing that it should be equivalent with PICASVM (i.e. we can think about removing that class which I left in for backwards compat).

File: pica/ml/classifiers/svm.py [hunk 3]
```diff
@@ -4,29 +4,114 @@
...
+        weights = {feature: mean_weights[i] for feature, i in names}
+
+        # sort by absolute value
+        sorted_weights = {feature: weights[feature] for feature in sorted(weights, key=lambda key: abs(weights[key]),
+                                                                          reverse=True)}
+        # TODO: weights should be adjusted if multiple original features were grouped together. probably not needed
+        #  if we rely on feature selection in near future
+        return sorted_weights
+
 class PICASVM:
```
Comment (phyden) [inline pica/ml/classifiers/svm.py:115]: Are we keeping this old class only for the `run_pica.py ` script? I just tested, we can also simply replace PICASVM for TrexSVM in the script - and maybe also add a parameter to toggle XGB instead of LinearSVM.

Comment (LokiLuciferase) [reply pica/ml/classifiers/svm.py:115]: I only kept PICASVM because I didn't want to break any existing workflows. As far as I see, TrexSVM is a drop-in replacement of PICASVM. So I am fine with replacing it!

Comment (phyden) [reply pica/ml/classifiers/svm.py:115]: Right. For me it would not break anything, and as far as I know I'm the only user of trex at the moment. Others we let speak now or remain silent forever :smiley:

Comment (VarIr) [reply pica/ml/classifiers/svm.py:115]: Silentium

Comment (LokiLuciferase) [reply pica/ml/classifiers/svm.py:115]: just merge it whenever you feel like it :)

**Review by VarIr — APPROVED** [review]

**Comment by VarIr** [comment]
We had a spurious fail on Travis with MacOS for the dev branch. I restarted the Travis job, and now it passed. Don't know what happened here, but just keep in mind, that these kinds of problems may occur, and may be very simple to solve.
